### PR TITLE
feat(facets-next): add `HierarchicalFilterEntity`

### PR DESCRIPTION
**Pull request by javieri-empathy**
# Description

Adds the `HierarchicalFilterEntity` following the task diagram.

# Context

This is needed as part of the new facets module.

# Test

Check that the hierarchical filter entity works as expected, supporting multiselection by default.

### CODE REVIEW

**Review by pmareke — APPROVED** [review]
LGTM

File: packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Saves the given filter to the store.
+   *
+   * @param filter - The filter to save to the store.
+   * @internal
+   */
+  protected saveFilter(filter: HierarchicalFilter): void {
+    this.store.commit('x/facetsNext/setFilter', filter);
+  }
+}
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:73]: Maybe create a method to cast only once?
```suggestion

  protected getFilterById(id: HierarchicalFilter['id']): HierarchicalFilter {
    return this.store.state.x.facetsNext.filters[id] as HierarchicalFilter;
  }
}
```

File: packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   * Deselects all the descendants of the given filter, saving them to the store.
+   *
+   * @remarks The descendants hierarchy is retrieved from the store.
+   * @param filter - The filter to deselect its descendants.
+   * @internal
+   */
+  protected deselectDescendants(filter: HierarchicalFilter): void {
+    if (filter.children) {
+      filter.children.forEach(childId => {
+        const child = this.store.state.x.facetsNext.filters[childId] as HierarchicalFilter;
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:57]:
```suggestion
        const child = this.getFilterById(childId);
```

File: packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Selects all the ancestors of the given filter, saving them to the store.
+   *
+   * @remarks The ancestors hierarchy is retrieved from the store.
+   * @param filter - The filter to select its ancestors.
+   * @internal
+   */
+  protected selectAncestors(filter: HierarchicalFilter): void {
+    if (filter.parentId) {
+      const parent = this.store.state.x.facetsNext.filters[filter.parentId] as HierarchicalFilter;
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:41]:
```suggestion
      const parent = this.getFilterById(filter.parentId);
```

**Review by tajespasarela — CHANGES_REQUESTED** [review]

**Review by tajespasarela — DISMISSED** [review]

File: packages/x-components/src/__stubs__/filters-stubs.factory.ts [hunk 4]
```diff
@@ -375,3 +375,49 @@ export function createNextEditableNumberRangeFilter(
...
+
+/**
+ * Creates a factory of
+ * {@link @empathyco/x-types-next#HierarchicalFilter | HierarchicalFilter}s.
+ *
+ * @param facetId - Facet id to which the filter belongs.
+ * @param parentId - Filter's parent id if exists.
+ * @returns A scoped function which is able to create `HierarchicalFilters`.
+ */
+export function createNextHierarchicalFilterFactory(
```
Comment (tajespasarela) [inline packages/x-components/src/__stubs__/filters-stubs.factory.ts:398]: Can we use this factory to define the `createNextHierarchicalFilter` function above?

File: packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+   * Saves the given filter to the store.
+   *
+   * @param filter - The filter to save to the store.
+   * @internal
+   */
+  protected saveFilter(filter: HierarchicalFilter): void {
+    this.store.commit('x/facetsNext/setFilter', filter);
+  }
+
+  protected getFilterById(id: HierarchicalFilter['id']): HierarchicalFilter {
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:80]: documentation here?

File: packages/x-components/src/__stubs__/filters-stubs.factory.ts [hunk 6]
```diff
@@ -375,3 +375,49 @@ export function createNextEditableNumberRangeFilter(
...
+      : [];
+    return [
+      {
+        id: filterId,
+        facetId: facetId,
+        parentId,
+        selected,
+        label,
+        totalResults: 10,
+        callbackInfo: {},
```
Comment (tajespasarela) [inline packages/x-components/src/__stubs__/filters-stubs.factory.ts:415]: remove the `callbackInfo` field

File: packages/x-components/src/__stubs__/filters-stubs.factory.ts [hunk 7]
```diff
@@ -375,3 +375,49 @@ export function createNextEditableNumberRangeFilter(
...
+ *
+ * @param facetId - Facet id to which the filter belongs.
+ * @param parentId - Filter's parent id if exists.
+ * @returns A scoped function which is able to create `HierarchicalFilters`.
+ */
+export function createNextHierarchicalFilterFactory(
+  facetId: string,
+  parentId: string | null = null
+): CreateNextHierarchicalFilter {
+  return (label, selected, createChildren) => {
```
Comment (tajespasarela) [inline packages/x-components/src/__stubs__/filters-stubs.factory.ts:402]: If you had added the type, then you would realize about the `callbackInfo` and `value` fields
```suggestion
  return (label, selected, createChildren): NextHierarchicalFilter[] => {
```

File: packages/x-components/src/__stubs__/filters-stubs.factory.ts [hunk 8]
```diff
@@ -375,3 +375,49 @@ export function createNextEditableNumberRangeFilter(
...
+      {
+        id: filterId,
+        facetId: facetId,
+        parentId,
+        selected,
+        label,
+        totalResults: 10,
+        callbackInfo: {},
+        children: children.map(filter => filter.id),
+        value: label.toLowerCase().replace(/\s+/g, '-'),
```
Comment (tajespasarela) [inline packages/x-components/src/__stubs__/filters-stubs.factory.ts:417]: remove the `value` field

File: packages/x-components/src/x-modules/facets-next/entities/__tests__/utils.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+/**
+ * Creates a store with the {@link facetsNextXStoreModule} registered, and initialises its
+ * state with the provided filters.
+ *
+ * @param filters - The filters to set to the {@link facetsNextXStoreModule} module state.
+ * @returns A store with the {@link facetsNextXStoreModule} registered, and the provided facets.
+ */
+export function prepareFacetsStoreWithFilters(filters: Filter[]): Store<RootXStoreState> {
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets-next/entities/__tests__/utils.ts:36]: what about using only one function `prepareFacetsStore` with optional `filters: Filter[]` parameter?

**Review by tajespasarela — CHANGES_REQUESTED** [review]

File: packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts (hunk 5 above)
Comment (javieri-empathy) [reply packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:80]: That is what happens when you accept suggestions in the PR. No more github suggestions, manual copy and paste ftw

File: packages/x-components/src/__stubs__/filters-stubs.factory.ts (hunk 7 above)
Comment (javieri-empathy) [reply packages/x-components/src/__stubs__/filters-stubs.factory.ts:402]: I didn't know it was using relaxed types when it inferred return types. Good catch

File: packages/x-components/src/__stubs__/filters-stubs.factory.ts (hunk 4 above)
Comment (javieri-empathy) [reply packages/x-components/src/__stubs__/filters-stubs.factory.ts:398]: We tried, but I wouldn't: this factory creates a hierarchy of filters, and returns them flattened, while the other method, `createNextHierarchicalFilter` only creates a single filter and returns it. They have different purposes.

**Review by tajespasarela — APPROVED** [review]

**Review by pmareke — APPROVED** [review]

**Pull request merged**